File: android/app/src/main/kotlin/com/farosos/app/ConnectivityMonitor.kt
package com.farosos.app

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest

/**
 * Envoltorio de `ConnectivityManager.NetworkCallback`, sin interfaz —
 * mismo molde que `BleAdvertiser`/`BleScanner`. Sin ping ni request
 * propio: el filtro exige `NET_CAPABILITY_INTERNET` +
 * `NET_CAPABILITY_VALIDATED`, que el sistema ya resuelve por su cuenta —
 * a diferencia de iOS (`NWPathMonitor`, best-effort), esto sí confirma
 * salida real a internet.
 */
class ConnectivityMonitor(private val context: Context) {
    var onConnectivityChanged: ((Boolean) -> Unit)? = null

    private val connectivityManager
        get() = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    private val request = NetworkRequest.Builder()
        .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
        .addCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
        .build()

    private val callback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            onConnectivityChanged?.invoke(true)
        }

        override fun onLost(network: Network) {
            onConnectivityChanged?.invoke(false)
        }
    }

    fun start() {
        connectivityManager.registerNetworkCallback(request, callback)
    }

    /** Desregistra el callback — se llama al destruirse el `ViewModel`. */
    fun stop() {
        connectivityManager.unregisterNetworkCallback(callback)
    }
}

File: ios/EmergencyApp/Sources/ConnectivityMonitor.swift
import Foundation
import Network

/// Envoltorio de `NWPathMonitor`, sin protocolo/interfaz — mismo molde que
/// `BleAdvertiser`/`BleScanner`. Sin ping ni request propio: usa
/// exclusivamente la validación pasiva del sistema operativo.
///
/// Best-effort en esta plataforma (ticket #17, asimetría documentada): a
/// diferencia de Android (`NET_CAPABILITY_VALIDATED`, que sí confirma
/// salida real a internet), `path.status == .satisfied` solo confirma que
/// hay una interfaz con ruta activa — un WiFi con portal cautivo puede
/// reportar `.satisfied` sin tener internet real.
final class ConnectivityMonitor {
    var onConnectivityChanged: ((Bool) -> Void)?

    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "com.farosos.connectivitymonitor")

    func start() {
        monitor.pathUpdateHandler = { [weak self] path in
            let hasConnectivity = path.status == .satisfied
            DispatchQueue.main.async {
                self?.onConnectivityChanged?(hasConnectivity)
            }
        }
        monitor.start(queue: queue)
    }
}
